crypto/bn/bn_exp.c

```diff
@@ -618,6 +618,15 @@ int BN_mod_exp_mont_consttime(BIGNUM *rr, const BIGNUM *a, const BIGNUM *p,
 
     top = m->top;
 
+    if (in_mont != NULL && BN_is_zero(&in_mont->N)) {
+        BNerr(BN_F_BN_MOD_EXP_MONT_CONSTTIME, ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+    if ((unsigned int)top > INT_MAX / sizeof(m->d[0]) / (1 << 8)) {
+        /* Prevent overflowing the powerbufLen computation below */
```

Comment:
have you seen the `np = alloca(top * sizeof(BN_ULONG));` down below ?

Reply:
heh, no. Now that is SPARC only code. Not sure what to do with that. Change it to OPENSSL_malloc() for some value of top? Can be a separate PR.

Reply:
Could we solve this differently, like when the modulus size is larger than RSA_MAX_MODULUS then ignore the constant time flag, since that large numbers are usually just used for denial of service attacks?

Reply:
I mean the non-constant time code will be orders of magnitude faster than this, and will not have these alloca problem.

Reply:
If you're concerned about the `alloca`, note that `bn_mul_mont`  itself, used also by `BN_mod_exp_mont`, internally does a similar `alloca`, hidden away in assembly. It may be worth just cutting down the largest BIGNUM a little. (I'll probably go that route in BoringSSL.)

If we do that, this actually won't be able to overflow. The largest BIGNUM is capped here. It's `INT_MAX / (4 * BN_BITS2)` words, or `INT_MAX / 4` bits, or `INT_MAX / 32` bytes, or 64MiB.
https://github.com/openssl/openssl/blob/be0161ff100bf10c9549fc09ce4513681011da1c/crypto/bn/bn_lib.c#L268

Scrolling down, `powerbufLen` increases from two places, a `top * sizeof(mont->N.d[0])` term (which, if it happens, forces `window` to 5), and the more complex term:

```
    powerbufLen += sizeof(m->d[0]) * (top * numPowers +
                                      ((2 * top) >
                                       numPowers ? (2 * top) : numPowers));
```

(Incidentally, I believe the `(2 * top) > numPowers ? (2 * top) : numPowers` bit is a remnant of older versions of mont5 which used to overread the table by one row. Unreasonably clever scheduling. It no longer does this and probably can be replaced with `2 * top` now, though I haven't carefully analyzed all the files in OpenSSL, just the subset we have.)

`numPowers` is at most 64, so we're looking at 64 BIGNUMs, plus change, or a bit over 4GiB. That won't fit in int, but it's close enough that we don't need to bring it down too much further to make it always fit.

Reply:
>  It may be worth just cutting down the largest BIGNUM a little. (I'll probably go that route in BoringSSL.)

Update: this _almost_ works, except we found one project that relies on making 8MiB bignums, and some others that needed 32-64KiB. 32-64KiB is plausible to alloca, but 8MiB is too much. So we're probably going to apply the cap Montgomery moduli instead.

Reply:
Hmm, isn't a kernel task stack just 64 K large?

Reply:
Hmm yeah that's possibly also a bit large. Regardless, the 8MiB one means a bn-wide limit doesn't seem viable. My plan was to do a 8KiB limit in BN_MONT_CTX, which is enough for the maximum of all our cryptographic primitives, with a little additional leeway. (The callers that needed 32-64KiB also don't use Montgomery reduction.)

Reply:
Take a look at #19735 is that acceptable, or did I miss someting?